### PR TITLE
Add initial attempt at an ssh plugin

### PR DESCRIPTION
**Pull request by jpevarnek**
First of all, full credit for the build work here goes to @Shanjian

This adds a basic SSH plugin that will work with username/password auth.
Calling connect gets you a connection object that then has methods for
either executing a command or disconnecting.

Coming soon: auth by key

[Review on Reviewable](https://reviewable.io/reviews/uproxy/colony/16)

### CODE REVIEW

File: client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  @Override
+  public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
+    try {
+      if (action.equals("connectPassword")) {
+        String host = args.getString(0);
+        int port = args.getInt(1);
+        String username = args.getString(2);
+        String password = args.getString(3);
+
+        this.connect(host, port, username, password, callbackContext);
```
Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java:45]: Can we just make this consistent? "connectPassword"

Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java:45]: Nvm. I see what you're doing below. I'm fine with this

File: client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    if (password != null) {
+      session.setPassword(password);
+    }
+
+    // Avoid asking for key confirmation
+    Properties prop = new Properties();
+    prop.put("StrictHostKeyChecking", "no"); //TODO this is bad
+    session.setConfig(prop);
+
+    session.connect();
```
Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java:90]: How are failures returned?

File: client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    this.connect(host, port, username, null, callbackContext);
+  }
+
+  private void sendCommand(int idConnection, String command, CallbackContext callbackContext) throws JSchException, IOException, JSONException {
+    Session session = this.getSession(idConnection);
+    ChannelExec channel = (ChannelExec)session.openChannel("exec");
+    channel.setCommand(command);
+
+    ByteArrayOutputStream output = new ByteArrayOutputStream();
```
Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java:108]: not used?

Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java:108]: http://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html

You can just write to the bytearrayoutputstream and then call toByteArray at the end to retrieve the full thing

File: client/plugin-src/cordova-plugin-ssh/ssh.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/* globals cordova, window */
+
+window.ssh = {};
+window.ssh._state = null;
+
+var sendCommand = function(method, params) {
```
Comment (ryscheng) [inline client/plugin-src/cordova-plugin-ssh/ssh.js:6]: Not super important, but I have a slightly more concise version of the same thing here:
https://github.com/uProxy/colony/blob/ryscheng-openvpn/client/plugin-src/cordova-plugin-openvpn/openvpn.js

File: gulpfile.js [hunk 5]
```diff
@@ -94,10 +96,19 @@ gulp.task("setup", gulpSequence(
...
   "cordova_plugin_oauthredirect",
+  "cordova_plugin_ssh",
   "setup_www",
-  "cordova_build",
-  "cordova_emulate"
+  "cordova_build"
+  // "cordova_emulate"
 ));
+
+gulp.task("ssh", gulpSequence(
```
Comment (ryscheng) [inline gulpfile.js:105]: Can we just sick to gulp setup / remove this mblock?

File: gulpfile.js [hunk 6]
```diff
@@ -94,10 +96,19 @@ gulp.task("setup", gulpSequence(
   "cordova_create",
   "cordova_platform_android",
   "cordova_plugin_oauthredirect",
+  "cordova_plugin_ssh",
   "setup_www",
-  "cordova_build",
-  "cordova_emulate"
+  "cordova_build"
+  // "cordova_emulate"
```
Comment (ryscheng) [inline gulpfile.js:102]: I agree, let's just delete this. No need to emulate after setup

**Comment by jpevarnek** [comment]
Review status: 0 of 161 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java, line 90 [r1]](https://reviewable.io:443/reviews/uproxy/colony/16#-JuNRaEHgAZvbAIr1hDG-r1-90)** ([raw file](https://github.com/uproxy/colony/blob/229aa119c36bc0a9a120bd4153e60383185012a4/client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java#L90)):</sup>
This throws an exception and it ends up getting switched to a rejected promise.

---

<sup>**[client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java, line 108 [r1]](https://reviewable.io:443/reviews/uproxy/colony/16#-JuNRaEHgAZvbAIr1hDG-r1-108)** ([raw file](https://github.com/uproxy/colony/blob/229aa119c36bc0a9a120bd4153e60383185012a4/client/plugin-src/cordova-plugin-ssh/src/android/Ssh.java#L108)):</sup>
Done.

---

<sup>**[client/plugin-src/cordova-plugin-ssh/ssh.js, line 6 [r1]](https://reviewable.io:443/reviews/uproxy/colony/16#-JuNRaETIpQWJmflHfwR-r1-6)** ([raw file](https://github.com/uproxy/colony/blob/229aa119c36bc0a9a120bd4153e60383185012a4/client/plugin-src/cordova-plugin-ssh/ssh.js#L6)):</sup>
Done.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/colony/16)

**Pull request merged**